### PR TITLE
Introduce NumericInput and redesign Spinner widget

### PR DESCRIPTION
**Pull request by xavArtley**
Following the dicussion on https://github.com/bokeh/bokeh/issues/10179
I propose a first implementation of a `NumericInput` widget
From which I hope to derive a `Spinner` which will not depend of the browser implementation

I'll tag it as a WIP, feel free to add your suggestions

### CODE REVIEW

File: bokehjs/src/lib/models/widgets/numeric_spinner.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    if (document.activeElement !== this.input_el) return
+    evt.preventDefault()
+
+    const sign = evt.deltaY > 0 ? -1 : 1
+    this.increment(sign * this.model.step)
+  }
+
+  _input_key_down(evt: KeyboardEvent) {
+    switch (evt.keyCode) {
+      case 38: // arrow up
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/numeric_spinner.ts:112]: `core/dom` has `Keys` enum to help with this. If a value is missing, update the enum and use it here.

File: bokehjs/src/lib/models/widgets/numeric_spinner.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+  increment(step: number): void {
+    const {low, high} = this.model
+    if (this.model.value == null) {
+      if (step > 0)
+        this.model.value =
+          low != null ? low : high != null ? min(0, high) : 0
+      else if (step < 0)
+        this.model.value =
+          high != null ? high : low != null ? max(low, 0) : 0
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/numeric_spinner.ts:139]: In bokehjs we don't split up lines like this. Longer lines are perfectly fine. I would also use parentheses around nested ternary expressions, to aid readablility.

**Comment by bryevdv** [comment]
@xavArtley I think the only thing failing the tests is a JS/Python default mismatch:
```
    Spinner.wheel_wait: bokehjs defaults to 100 but python defaults to 200
```

**Comment by xavArtley** [comment]
Hello @bryevdv , I have to finish this PR but I'm pretty busy these days.

**Comment by bryevdv** [comment]
@xavArtley No worries I was just pointing out what was the issue for reference! FYI our next release is tentatively around the ned of the month but if not by then we are planning to have minor releases on a 4-6 week cadence so this can go in any time at your pace in still see a release shortly after.

**Comment by bryevdv** [comment]
@xavArtley @mattpap @philippjfr I can just edit the files on the GH to get this past the finish line. But the question is which value is the correct default to use? 100ms or 200ms?

**Comment by philippjfr** [comment]
Hard to say without testing but looks like @xavArtley just pushed up a change to 100 ms. I'm happy deferring to his judgement.

**Comment by xavArtley** [comment]
It really depend of the reactiveness wanted of the application and by rule
of thumb

Le jeu. 30 juil. 2020 à 14:25, Philipp Rudiger <notifications@github.com> a
écrit :

**Comment by bryevdv** [comment]
@xavArtley there was a recent chrome change that broke our integration tests, this has been fixed but this PR will need to be  rebased against branch-2.2 to start passing tests.  However I also notice this PR targets master and we have also changed our branching strategy. This will need to land on branch-2.2 Can you either edit the PR branch, and re-base it on current branch-2.2? Or re-submit a new PR against current branch-2.2 ? That should be the last thing I promise :)

**Comment by bryevdv** [comment]
I just want to play around a bit with this locally, and hopefully merge later today.

**Comment by bryevdv** [comment]
Thanks @xavArtley !

**Comment by mattpap** [comment]
@bryevdv, tests were not run on this PR.

**Comment by bryevdv** [comment]
It listed green, what happened?

**Comment by bryevdv** [comment]
I pushed the linter fixes in https://github.com/bokeh/bokeh/commit/93666af4e6f17f1451933f1381ec0d3683583420 All the integration tests are passing for me (and on CI AFAICT)

**Pull request merged**